app/models/test_track/session.rb

```diff
@@ -52,41 +43,49 @@ def state_hash
   end
 
   def log_in!(identity, forget_current_visitor: nil)
-    identifier_type = identity.test_track_identifier_type
-    identifier_value = identity.test_track_identifier_value
-
-    @visitor = TestTrack::Visitor.new if forget_current_visitor
-    visitor.link_identifier!(identifier_type, identifier_value)
+    @unauthenticated_visitor = TestTrack::Visitor.new if forget_current_visitor
+    unauthenticated_visitor.link_identity!(identity)
+    visitors_by_identity[identity] = unauthenticated_visitor
 
-    identities << identity if identity.present?
     true
   end
 
   def sign_up!(identity)
-    identifier_type = identity.test_track_identifier_type
-    identifier_value = identity.test_track_identifier_value
-
-    visitor.link_identifier!(identifier_type, identifier_value)
-    identities << identity if identity.present?
+    unauthenticated_visitor.link_identity!(identity)
+    visitors_by_identity[identity] = unauthenticated_visitor
 
     TestTrack.analytics.sign_up!(visitor.id)
 
     true
   end
 
-  def has_matching_identity?(identity)
-    identities.include?(identity)
-  end
-
   private
 
   attr_reader :controller
 
   def visitor
-    @visitor ||= TestTrack::Visitor.new(id: visitor_id)
+    if current_identity
+      visitors_by_identity[current_identity]
+    else
+      unauthenticated_visitor
+    end
+  end
+
+  def current_identity
+    raise <<~ERROR unless controller.class.test_track_identity&.is_a?(Symbol)
+      Your controller (or controller base class) must set test_track_identity for
+      TestTrack to work properly. e.g.:
+
+        self.test_track_identity = :current_user
+    ERROR
```

Comment:
If we want to avoid the requirement of specifying an empty method, we could document a special symbol :none that we handle here as nil automatically.

It’s a small thing that would provide a nicer experience for anyone that does happen to use this in a system that doesn’t have auth

Reply:
Yeah, I spose nobody’s got a signed-in-entity getter called `#none`